File: apps/lite-web/src/eco/data/site.ts
import { styleById } from "./homepage";
import { fieldById } from "./tree";
import { READINGS, STUDENT, WRITINGS } from "./library";
import { motiveOf } from "./cards";
import { ARTIFACTS } from "./artifacts";
import type { HomepageSection, Project, StyleId } from "./types";

/**
 * 她做出来的那个网站 — content.
 *
 * ## Two rules this file exists to hold
 *
 * **① It is HER site, not our product.** Nothing on this page may be written in
 * the app's voice. No 「我做过的」/「我写的」 rubric over every block, no
 * 「为什么做它：」 field label, no footer sentence explaining our pedagogy. Real
 * personal sites are written by a person to strangers, and a person does not
 * label her own paragraphs. The section names are the ones people actually use
 * — 项目 · 文章 · 关于 · 在读 — and each appears **once**.
 *
 * **② Every post gets a written line, never a truncated one.** Slicing 44
 * characters off the body and adding 「…」 is how a scraper presents someone
 * else's writing. A blurb is a sentence the author chose.
 *
 * ## Where the content comes from
 * Precedence, most-hers first: what she rewrote in 内容草稿 (`hp-content`
 * blocks — 印记 promised those ship 一个字不动) → what she typed in the studio
 * → the mock below. Projects, posts and reading come from real state.
 */

export interface SiteTheme {
  paper: string;
  ink: string;
  accent: string;
  font: string;
}

export interface SiteProject {
  id: string;
  year: string;
  kind: string;
  title: string;
  /** One or two sentences, in her voice, with the reason folded into the
   *  prose. On a real site the reason IS the description. */
  blurb: string;
  /** The two-colour wash that stands in for a photograph. */
  plate: [string, string];
}

export interface SitePost {
  id: string;
  date: string;
  title: string;
  blurb: string;
  /** 分类 — the meta row every blog post list has. */
  kind: string;
  words: number;
  tags: string[];
}

export interface SiteRead {
  id: string;
  title: string;
  source: string;
  takeaway: string;
}

export interface SiteContent {
  name: string;
  handle: string;
  domain: string;
  /** The one-line status. A real convention (the /now page), not a widget. */
  now: string;
  /** The sentence the whole page is built around. */
  headline: string;
  lead: string;
  /** 一句话身份, under the name. Every reference site has this line. */
  role: string;
  /** The spaced-out line under a masthead: 算法 · 工程 · 产品 · 日常. */
  motto: string[];
  /** 站点信息 / colophon rows — label, value. */
  stats: { label: string; value: string }[];
  /** 标签云. */
  tags: string[];
  projects: SiteProject[];
  posts: SitePost[];
  reads: SiteRead[];
  about: string[];
  detail: string;
  nowList: string[];
  email: string;
  updated: string;
}

/* ── the mock ─────────────────────────────────────────────────────────── */

const BASE: SiteContent = {
  name: STUDENT.name,
  handle: STUDENT.handle,
  domain: "zhiyao.me",
  now: "在给小区花园画一张不会迷路的地图",
  role: "初二学生 · 在拆东西 · 在写为什么它们修不好",
  motto: ["拆解", "修理", "设计", "日常"],
  stats: [
    { label: "建站", value: "2026 年 3 月" },
    { label: "写了", value: "4 篇" },
    { label: "做了", value: "3 件" },
    { label: "更新", value: "大约两周一次" },
  ],
  tags: ["修理权", "拆解", "设计", "例外与代表性", "让步段", "螺丝"],
  headline: "一件还能修的东西，是谁决定它该被扔的？",
  lead: "我十四岁，在拆家里所有还能拆的东西。这一页放我做过的、写过的，和我最近在想的问题。",
  projects: [
    {
      id: "s-pill",
      year: "2026",
      kind: "做的东西",
      title: "让爷爷看得清的药盒",
      blurb:
        "他上周吃错了一次药。我做了三版：第一版好看但他看不清，第二版装不下他的药，第三版用红胶带分格，最丑，但他现在自己会拿。",
      plate: ["#6E4F8E", "#33224A"],
    },
    {
      id: "s-garden",
      year: "2026",
      kind: "在做",
      title: "让人在花园里不迷路",
      blurb:
        "外婆在自己住了九年的小区里迷过一次路。十四个路口、一张手画的地图、一批贴在树上的编号，还差物业那一关。",
      plate: ["#3F7E5C", "#1C4432"],
    },
    {
      id: "s-lamp",
      year: "2025",
      kind: "拆解",
      title: "一盏修不好的台灯",
      blurb:
        "拆到第七颗螺丝，才看见那块板子是焊死的。售后说整只换，或者扔。这一页上所有东西都是从那天开始的。",
      plate: ["#4A4E60", "#22242E"],
    },
  ],
  posts: [],
  reads: [],
  about: [
    "我十四岁，读初二。三年前家里一盏台灯坏了，我拆开看，里面是一块焊死的电路板，售后说只能整只换。那天我第一次意识到，「修不好」有时候不是坏得太狠，是被设计成这样的。",
    "从那以后我做的东西基本都在回答同一个问题：一件还能用的东西，是谁决定它该被扔的。我做过一个我爷爷能看清的药盒，写过一篇关于修理权的短文，现在在给我们小区的花园画一张地图。",
    "我留着每一个我拆开过的东西的螺丝，装在一个铁饼干盒里，现在有 213 颗。",
  ],
  detail: "我留着每一个我拆开过的东西的螺丝，装在一个铁饼干盒里，现在有 213 颗。",
  nowList: [
    "在数花园里的路口，数到第十四个。",
    "在读一篇关于一条河的报道，治理了十一年，只补回 3%。",
    "在试第三次修那盏台灯，这次买了热风枪。",
  ],
  email: "hi@zhiyao.me",
  updated: "2026-08-31",
};

/** 🚨 Written lines, one per post. See rule ② at the top of this file. */
const POST_BLURBS: Record<string, string> = {
  "w-coral": "红海那片珊瑚没有白化，是好消息。但它只有 4 平方公里。",
  "w-three-seconds": "我试着在别人说完之后数三下再开口，试了一周。",
  "w-repair": "修灯的师傅看了一眼说：这个胶封死了，拆开就废了。",
  "w-letters": "十万封家书里出现最多的三个词：钱、天气、你吃了吗。",
};

/* ── builders ─────────────────────────────────────────────────────────── */

function postsFrom(): SitePost[] {
  return [...WRITINGS]
    // 🚨 Newest first. A blog listing out of date order reads as broken before
    // a visitor has read a single word of it.
    .sort((a, b) => (a.date < b.date ? 1 : -1))
    .slice(0, 4)
    .map((w) => ({
      id: w.id,
      date: w.date,
      title: w.title,
      blurb: POST_BLURBS[w.id] ?? "",
      kind: fieldById(w.field).label,
      words: w.words,
      tags: w.keywords,
    }));
}

function readsFrom(): SiteRead[] {
  return READINGS.filter((r) => r.takeaway)
    .slice(0, 3)
    .map((r) => ({
      id: r.id,
      title: r.title,
      source: r.source,
      takeaway: r.takeaway ?? "",
    }));
}

/** Muted two-colour washes. Deliberately NOT the app's cover gradients with a
 *  glyph on them — a tile with an icon in the middle is app furniture, and it
 *  is the single thing that made this page read as a product screen. */
const PLATES: [string, string][] = [
  ["#6E4F8E", "#33224A"],
  ["#3F7E5C", "#1C4432"],
  ["#4A4E60", "#22242E"],
  ["#A2603C", "#4E2A18"],
  ["#3C6480", "#1B3242"],
  ["#8A5566", "#42222E"],
];

function projectsFrom(projects: Project[]): SiteProject[] {
  // 🚨 The site does not list itself. The 个人主页 project's deliverable IS this
  // page, so a card for it points a visitor at the page they are reading.
  const done = projects.filter((p) => p.phase === "published" && p.track !== "website");
  if (done.length === 0) return BASE.projects;
  return done.map((p, i) => {
    const why = motiveOf(p)?.cost ?? "";
    const summary = (p.summary ?? p.intent).trim();
    return {
      id: p.id,
      year: p.startedAt.slice(0, 4),
      kind: p.phase === "published" ? "做的东西" : "在做",
      title: p.title,
      // The reason leads, because that is how a person introduces their own
      // work — never as a labelled field underneath it.
      blurb: why ? `${why}${summary}` : summary,
      plate: PLATES[i % PLATES.length] ?? PLATES[0]!,
    };
  });
}

/** Her rewrites of 印记's draft, keyed by block id. */
function draftBlocks(projects: Project[]): Record<string, string> {
  const spec = ARTIFACTS["hp-content"];
  const out: Record<string, string> = {};
  if (!spec?.blocks) return out;
  const site = projects.find((p) => p.artifacts["hp-content"]);
  const edits = site?.artifacts["hp-content"]?.blocks ?? {};
  for (const b of spec.blocks) {
    const v = (edits[b.id] ?? "").trim();
    if (v) out[b.id] = v;
  }
  return out;
}

export function buildSite(args: {
  projects: Project[];
  sections?: HomepageSection[];
}): SiteContent {
  const { projects, sections = [] } = args;
  const blocks = draftBlocks(projects);
  const studio = (id: string) =>
    (sections.find((s) => s.id === id && s.enabled)?.value ?? "").trim();
  const pick = (blockId: string, sectionId: string, fallback: string) =>
    blocks[blockId] || studio(sectionId) || fallback;

  const detail = pick("detail", "detail", BASE.detail);
  return {
    ...BASE,
    headline: pick("question", "question", BASE.headline),
    lead: pick("intro", "intro", BASE.lead),
    detail,
    // The odd detail is the last thing she says about herself, so it is the
    // last paragraph of 关于 — not a boxed callout with a label on it.
    about: [...BASE.about.slice(0, 2), detail],
    email: studio("contact") || BASE.email,
    projects: projectsFrom(projects),
    posts: postsFrom(),
    reads: readsFrom(),
  };
}

/* ── theme ────────────────────────────────────────────────────────────── */

const FALLBACK: SiteTheme = {
  paper: "#F7F3EA",
  ink: "#2A2724",
  accent: "#A6402C",
  font: '"Noto Serif SC","Songti SC",Georgia,serif',
};

/**
 * Which of the three 方案 she took. The id matters as much as the colours: the
 * three options describe three genuinely different pages (a long empty essay,
 * a dense index, a work-first spread), so `layoutOf` picks the layout and the
 * palette together. Swapping only the colours would make her decision — the
 * one she had to write a reason for — change nothing anyone can see.
 */
export function siteStyle(
  projects: Project[],
  style: StyleId | null,
): { layout: "essay" | "ledger" | "magazine"; theme: SiteTheme } {
  for (const p of projects) {
    const choice = p.artifacts["hp-style"]?.choice;
    const opt = ARTIFACTS["hp-style"]?.options?.find((o) => o.id === choice);
    if (opt) {
      return {
        layout: opt.id === "B" ? "ledger" : opt.id === "C" ? "magazine" : "essay",
        theme: {
          paper: opt.paper,
          ink: opt.ink,
          accent: opt.accent,
          font: opt.font ?? FALLBACK.font,
        },
      };
    }
  }
  if (style) {
    const s = styleById(style);
    return {
      layout: s.headline === "mono-caps" ? "ledger" : s.headline === "sans-tight" ? "magazine" : "essay",
      theme: { paper: s.paper, ink: s.ink, accent: s.accent, font: s.font },
    };
  }
  return { layout: "essay", theme: FALLBACK };
}

File: apps/api/internal/pbl/site.go
package pbl

// site.go — 她的主页的内容模型，以及「她自己写的字」和「她真做过的事」怎么合成
// 一个页面。
//
// ## 这个文件存在的理由
//
// 原型里这一层是三个写死的常量：`STUDENT = { name: "林知遥", handle: "zhiyao" }`、
// 一份 READINGS、一份 WRITINGS。跑一遍真实流程之后页面是这样的：页头写着
// 「林知遥 · 初二学生」，紧接着的一行是她自己写的「我是侯知遥，读 IB 的高二学
// 生」；关于那一段是别人的完整自传（台灯、螺丝、213 颗）；文章和作品两栏一条都
// 不是她的。十一个区块里只有两个是她的。
//
// 所以这里有一条硬规则：**页面上的每一个字，要么是她敲进去的，要么是从她真实的
// reading / writing / pbl_project 行里读出来的。这个文件里没有第三种来源。**
// 没有兜底文案，没有示例内容，没有「先放一段让页面好看点」。她没写的地方就是
// 空的，而空的地方由发布门槛（SiteMissing）挡住，不由假字填上。
//
// ## 为什么没有 handle 和 domain
//
// 原型给每个人发了一个 `zhiyao.me` 和 `hi@zhiyao.me`。那是一个不存在的域名和一
// 个不存在的信箱，印在页头和页脚上，是这个页面上最像真的一处假话。真实地址是
// 发布后拿到的那条 /p/<token> 链接，它在她自己的界面里显示，不假装成一个域名。
//
// ## 为什么联系方式默认是空的
//
// 不从账号邮箱自动带出来。spec §15 之所以让这条链接不可索引，就是因为她是未成
// 年人；把她的登录邮箱印在一个公开页面上，比让页面被搜到更糟。要不要留联系方
// 式、留哪个，是她自己决定的一件事。

import (
	"crypto/sha256"
	"encoding/hex"
	"fmt"
	"sort"
	"strings"
	"time"
)

// SiteTheme — 版式带的四个颜色/字体变量。渲染端只认这四个。
type SiteTheme struct {
	Paper  string `json:"paper"`
	Ink    string `json:"ink"`
	Accent string `json:"accent"`
	Font   string `json:"font"`
}

// SiteStat — 站点信息里的一行。
type SiteStat struct {
	Label string `json:"label"`
	Value string `json:"value"`
}

// SiteProject — 主页上的一件作品。
type SiteProject struct {
	ID    string `json:"id"`
	Year  string `json:"year"`
	Kind  string `json:"kind"`
	Title string `json:"title"`
	// 她自己写的那一两句。空着就是空着——真实个人站上，介绍作品的那句话就是
	// 作者写的那句话，截取正文前 44 个字再加省略号是爬虫干的事。
	Blurb string `json:"blurb"`
	// 代替照片的双色版。它不假装是一张照片，所以也不需要「这里缺一张照片」的
	// 虚线框——它就是这一页自己的图案。
	Plate [2]string `json:"plate"`
}

// SitePost — 主页上的一篇文章。
type SitePost struct {
	ID    string   `json:"id"`
	Date  string   `json:"date"`
	Title string   `json:"title"`
	Blurb string   `json:"blurb"`
	Kind  string   `json:"kind"`
	Words int      `json:"words"`
	Tags  []string `json:"tags"`
}

// SiteRead — 在读。
type SiteRead struct {
	ID     string `json:"id"`
	Title  string `json:"title"`
	Source string `json:"source"`
	// 她自己写的一句。我们不替她总结她读过的东西。
	Takeaway string `json:"takeaway"`
}

// SiteContent — 渲染端拿到的全部东西。三个版式都是它的纯函数。
type SiteSection struct {
	ImageKey string `json:"imageKey,omitempty"`
	ImageURL string `json:"imageUrl,omitempty"`
	Key      string `json:"key"`
	Title    string `json:"title"`
	Depth    int    `json:"depth"`
	Body     string `json:"body"`
}

type SiteContent struct {
	Sections []SiteSection `json:"sections,omitempty"`
	Name     string        `json:"name"`
	Role     string        `json:"role"`
	Headline string        `json:"headline"`
	Lead     string        `json:"lead"`
	Now      string        `json:"now"`
	Motto    []string      `json:"motto"`
	Stats    []SiteStat    `json:"stats"`
	Tags     []string      `json:"tags"`
	Projects []SiteProject `json:"projects"`
	Posts    []SitePost    `json:"posts"`
	Reads    []SiteRead    `json:"reads"`
	About    []string      `json:"about"`
	NowList  []string      `json:"nowList"`
	Email    string        `json:"email"`
	Updated  string        `json:"updated"`
	// 画头图用的种子。由她的名字派生，所以每个人的头图都不一样，而且同一个人
	// 每次都一样。见前端 parts.tsx 的 Banner。
	Seed int `json:"seed"`
}

// SiteDraft — 存进 pbl_site.content 的那一份，**只有她写的字**。
//
// 三张列表不在这里：它们每次渲染时从真实的行里现拼。存快照等于存一份会过期的
// 副本，而主页最不该做的事就是与她真做过的事对不上。
type SiteDraft struct {
	Sections []SiteSection `json:"sections,omitempty"`
	Role     string        `json:"role"`
	Headline string        `json:"headline"`
	Lead     string        `json:"lead"`
	Now      string        `json:"now"`
	Motto    []string      `json:"motto"`
	Tags     []string      `json:"tags"`
	About    []string      `json:"about"`
	NowList  []string      `json:"nowList"`
	Contact  string        `json:"contact"`
	// 每条作品 / 文章 / 在读，她自己写的那一句，按稳定 id 存。
	Blurbs map[string]string `json:"blurbs"`
}

// SiteItem — 从真实行里读出来的一条，三种之一。API 层负责把 sqlc 的行转成它，
// 这样这个文件不依赖 sqlc，可以纯逻辑地测。
type SiteItem struct {
	// AtomID 只用来算稳定 id，绝不进入返回给访客的内容。
	AtomID string
	Title  string
	When   time.Time
	// HasWhen 区分「零值时间」和「真的没有时间」——一篇没写完成时间的文章，
	// 日期栏该是空的，不该是 0001-01-01。
	HasWhen bool
	Words   int
	Source  string
	// Kind 是显示用的那个词：作品用 状态，文章用 分类。
	Kind string
}

/* ── 稳定 id ──────────────────────────────────────────────────────────── */

// SiteItemID 把一颗 atom 映射成主页上一个短的、不透明的 id。
//
// 🚨 不能直接用 atom 的 uuid。公开页面是任何人都能打开的，而 uuid 指向的是她
// 名下一行真实的数据，会出现在她其他链接里；把它印在公开 JSON 上，等于把一个
// 可寻址的句柄发给陌生人（atom_report_share.go 的文件头讲的是同一件事）。
//
// 也不能用下标（w1 / w2）。她给每条作品写的那句话是按 id 存的，而下标会在她写
// 完下一篇文章时整体移位——于是她为 A 写的介绍，某天早上出现在 B 底下。
//
// 所以：用户 id 加 atom id 一起哈希。同一个学生看同一条永远是同一个 id，换个
// 学生就完全不同，反推不回 atom id。
func SiteItemID(userID, atomID string) string {
	sum := sha256.Sum256([]byte(userID + ":" + atomID + ":site-item"))
	return hex.EncodeToString(sum[:6])
}

// siteSeed 由她的名字派生一个稳定的数，前端拿它画这一页自己的头图。
//
// 原型的头图是一盏台灯和散在天上的螺丝——那是林知遥的故事，画在每一个学生的
// 页顶上。头图得是这一页自己的图案，所以它由她派生，不由一个写死的场景派生。
func siteSeed(name string) int {
	sum := sha256.Sum256([]byte("banner:" + name))
	n := 0
	for _, b := range sum[:4] {
		n = n<<8 | int(b)
	}
	if n < 0 {
		n = -n
	}
	return n % 100000
}

/* ── 版面色板 ─────────────────────────────────────────────────────────── */

// sitePlates — 代替照片的双色版。刻意不是「app 的封面渐变 + 中间一个图标」：
// 中间带图标的方块是产品家具，也是原型那一版最像产品截图的一处。
var sitePlates = [][2]string{
	{"#6E4F8E", "#33224A"},
	{"#3F7E5C", "#1C4432"},
	{"#4A4E60", "#22242E"},
	{"#A2603C", "#4E2A18"},
	{"#3C6480", "#1B3242"},
	{"#8A5566", "#42222E"},
}

// plateFor 由稳定 id 选一块色版，所以同一件作品的颜色不会因为她新做了一个项目
// 而整排换掉。
func plateFor(id string) [2]string {
	n := 0
	for _, c := range id {
		n = (n*31 + int(c)) % 1000003
	}
	return sitePlates[n%len(sitePlates)]
}

/* ── 合成 ─────────────────────────────────────────────────────────────── */

// SiteInput 是 BuildSite 要的全部东西。
type SiteInput struct {
	// DisplayName 来自 users.display_name。这是她自己的名字，这一层没有别的
	// 名字可用，也不该有。
	DisplayName string
	Draft       SiteDraft
	Projects    []SiteItem
	Posts       []SiteItem
	Reads       []SiteItem
	// Since 建站时间；Updated 最后更新。
	Since   time.Time
	Updated time.Time
}

// BuildSite 把她写的字和她做过的事合成一页。
//
// 规则只有一条：这里不生成任何内容。它只搬运和排列。
func BuildSite(in SiteInput) SiteContent {
	d := in.Draft

	projects := make([]SiteProject, 0, len(in.Projects))
	for _, it := range in.Projects {
		projects = append(projects, SiteProject{
			ID:    it.AtomID,
			Year:  yearOf(it),
			Kind:  it.Kind,
			Title: it.Title,
			Blurb: strings.TrimSpace(d.Blurbs[it.AtomID]),
			Plate: plateFor(it.AtomID),
		})
	}

	posts := make([]SitePost, 0, len(in.Posts))
	for _, it := range in.Posts {
		posts = append(posts, SitePost{
			ID:    it.AtomID,
			Date:  dateOf(it),
			Title: it.Title,
			Blurb: strings.TrimSpace(d.Blurbs[it.AtomID]),
			Kind:  it.Kind,
			Words: it.Words,
			Tags:  nil,
		})
	}
	// 一份不按日期排的文章列表，读者还没读到第一个字就已经觉得这页是坏的。
	sort.SliceStable(posts, func(i, j int) bool { return posts[i].Date > posts[j].Date })

	reads := make([]SiteRead, 0, len(in.Reads))
	for _, it := range in.Reads {
		reads = append(reads, SiteRead{
			ID:       it.AtomID,
			Title:    it.Title,
			Source:   it.Source,
			Takeaway: strings.TrimSpace(d.Blurbs[it.AtomID]),
		})
	}

	stats := []SiteStat{}
	if !in.Since.IsZero() {
		stats = append(stats, SiteStat{Label: "建站", Value: in.Since.Format("2006 年 1 月")})
	}
	// 数字只报真的，而且为 0 的那一行直接不出现——「写了 0 篇」是一句没人会在
	// 自己主页上写的话。
	if n := len(posts); n > 0 {
		stats = append(stats, SiteStat{Label: "文章", Value: fmt.Sprintf("%d 篇", n)})
	}
	if n := len(projects); n > 0 {
		stats = append(stats, SiteStat{Label: "成果", Value: fmt.Sprintf("%d 件", n)})
	}
	if n := len(reads); n > 0 {
		stats = append(stats, SiteStat{Label: "阅读", Value: fmt.Sprintf("%d 篇", n)})
	}

	updated := ""
	if !in.Updated.IsZero() {
		updated = in.Updated.Format("2006-01-02")
	}

	return SiteContent{
		Sections: d.Sections,
		Name:     strings.TrimSpace(in.DisplayName),
		Role:     strings.TrimSpace(d.Role),
		Headline: strings.TrimSpace(d.Headline),
		Lead:     strings.TrimSpace(d.Lead),
		Now:      strings.TrimSpace(d.Now),
		Motto:    nonEmpty(d.Motto),
		Stats:    stats,
		Tags:     nonEmpty(d.Tags),
		Projects: projects,
		Posts:    posts,
		Reads:    reads,
		About:    nonEmpty(d.About),
		NowList:  nonEmpty(d.NowList),
		Email:    strings.TrimSpace(d.Contact),
		Updated:  updated,
		Seed:     siteSeed(in.DisplayName),
	}
}

func yearOf(it SiteItem) string {
	if !it.HasWhen {
		return ""
	}
	return it.When.Format("2006")
}

func dateOf(it SiteItem) string {
	if !it.HasWhen {
		return ""
	}
	return it.When.Format("2006-01-02")
}

func nonEmpty(xs []string) []string {
	out := make([]string, 0, len(xs))
	for _, x := range xs {
		if s := strings.TrimSpace(x); s != "" {
			out = append(out, s)
		}
	}
	return out
}

/* ── 发布门槛 ─────────────────────────────────────────────────────────── */

// SiteMissing 报告这一页还缺哪些**她自己的字**，按发布时该先补哪一个排序。
//
// 🚨 这是原型那个 bug 真正的解药。原型能发布一个一个字都不是她的页面，因为缺的
// 地方全被示例内容填满了，看上去是满的。这里没有示例内容，所以缺的地方是空的；
// 空的地方拦住发布，并且明说缺什么。
//
// 门槛只管三样：她是谁（role）、这一页在问什么（headline）、她想说的那段话
// （about）。作品和文章可以一条都没有——一个刚开始的人本来就没有；但一个没有
// 一句她自己的话的页面，发布出去代表不了任何人。
func SiteMissing(c SiteContent) []string {
	var out []string
	if c.Name == "" {
		out = append(out, "你的名字")
	}
	if c.Headline == "" {
		out = append(out, "首屏介绍")
	}
	if c.Role == "" {
		out = append(out, "个人简介")
	}
	if len(c.Sections) > 0 {
		filled := false
		for _, section := range c.Sections {
			if strings.TrimSpace(section.Body) != "" {
				filled = true
			}
		}
		if !filled {
			out = append(out, "主页模块的正文")
		}
	}
	// A confirmed outline holds the student's own introduction/content in
	// sections. Do not require a duplicate paragraph above that outline.
	if len(c.Sections) == 0 && len(c.About) == 0 {
		out = append(out, "关于我的正文")
	}
	return out
}

// SiteLayouts 是三个真正不同的版面（spec §15）。
var SiteLayouts = []string{"essay", "ledger", "magazine"}

// IsSiteLayout 校验她挑的那一个。
func IsSiteLayout(s string) bool {
	for _, l := range SiteLayouts {
		if l == s {
			return true
		}
	}
	return false
}

// SitePublishMissing is shared by the preview and publishing endpoint.
func SitePublishMissing(content SiteContent, palette Palette) []string {
	missing := SiteMissing(content)
	if !ValidPalette(palette) {
		missing = append(missing, "请在视觉基调中确认配色和风格")
	}
	if missing == nil {
		return []string{}
	}
	return missing
}
